### PR TITLE
prevent names from containing characters that break urls

### PR DESCRIPTION
**Pull request by deads2k**
API object names are used as part of the URL.  Since Go has trouble with URL escaping (see https://play.golang.org/p/EaCgpLYJLf, thanks @liggitt), we need to prevent objects from being created with those names.  The objects become unreferencable, so they can't be directly retrieved or deleted.

@ncdc fyi

### CODE REVIEW

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -81,6 +81,26 @@ func ValidateAnnotations(annotations map[string]string, field string) errs.Valid
...
+// Go doesn't properly handle url encoding.  Try:
+// 	u, _ := url.Parse("https://www.example.com/test%2Fpath/")
+// 	fmt.Println(u.Path)
+func ValidateMinimalName(name string, field string) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+
+	if len(name) == 0 {
+		allErrs = append(allErrs, errs.NewFieldRequired(field))
+	}
+	if strings.Contains(name, "%") {
```
Comment (brendandburns) [inline pkg/api/validation/validation.go]: turn this into a for loop?

```go
for _, str := range []string{"%", "/"} {
  ...
}
```

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -194,9 +214,8 @@ func ValidateObjectMeta(meta *api.ObjectMeta, requiresNamespace bool, nameFn Val
 	}
 	// if the generated name validates, but the calculated value does not, it's a problem with generation, and we
 	// report it here. This may confuse users, but indicates a programming bug and still must be validated.
-	if len(meta.Name) == 0 {
-		allErrs = append(allErrs, errs.NewFieldRequired("name"))
-	} else {
+	allErrs = append(allErrs, ValidateMinimalName(meta.Name, "name")...)
+	if nameFn != nil && (len(meta.Name) > 0) {
```
Comment (brendandburns) [inline pkg/api/validation/validation.go]: this ```nil``` check is irrelevant here b/c ```nameFn``` is used above on ```211``` (feel free to move it up)

**Comment by fabioy** [comment]
@bgrant0607 - Please reassign as appropriate.

@deads2k - Not sure why CLAbot is flagging you, but please resolve it. And rebase it again. Thank you.

File: pkg/api/validation/validation.go (hunk 2 above)
Comment (thockin) [inline pkg/api/validation/validation.go]: checking for length here is redundant with your minimal name.  Why not just inline the validate minimal function and make this simpler?

File: pkg/api/validation/validation.go [hunk 3]
```diff
@@ -1161,7 +1180,7 @@ func ValidateSecret(secret *api.Secret) errs.ValidationErrorList {
 
 func validateBasicResource(quantity resource.Quantity) errs.ValidationErrorList {
 	if quantity.Value() < 0 {
-		return errs.ValidationErrorList{fmt.Errorf("%v is not a valid resource quantity", quantity.Value())}
+		return errs.ValidationErrorList{errs.NewFieldInvalid("value", quantity.Value(), fmt.Sprintf("%v is not a valid resource quantity", quantity.Value()))}
```
Comment (thockin) [inline pkg/api/validation/validation.go]: merge issue?  Rohit should have fixed this earlier today..  @rjnagal did we miss one?

**Comment by thockin** [comment]
I'm OK with this, I guess, but none of our names should allow slashes and percents anyway - is there a real issue here?

**Comment by smarterclayton** [comment]
I question the premise of this - why is Go's library a reason for of potential future resources to not contain slashes or percents?  Why isn't this just part of certain types?

**Comment by bgrant0607** [comment]
How was this problem discovered?

**Comment by deads2k** [comment]
> How was this problem discovered?

One of our resources didn't think that it had any validation rules for names, because the resource didn't care what the name was.  Using this library, someone created a resource named "openshift/hello-world" and he was left unable to get, update, or delete the object.  Even the namespace finalizers couldn't remove it.

> why is Go's library a reason for of potential future resources to not contain slashes or percents? Why isn't this just part of certain types?

Creating are resource that we can't get, update, or delete and leaving a namespace permanently muddied is really bad.  I don't see why we'd ever want our generic etcd registry to allow such a thing to happen.

**Comment by deads2k** [comment]
> Not sure why CLAbot is flagging you, but please resolve it.

@smarterclayton Did we ever get the Red Hat CLA worked out?  Is there something special I need to do?

**Comment by smarterclayton** [comment]
@fabioy deads2k is covered by the Red Hat CLA - the bot just doesn't reflect Red Hat correctly right now.

@bgrant0607 see https://github.com/GoogleCloudPlatform/kubernetes/pull/734#discussion-diff-15708916 which was the original context for %2F.  Go is broken with respect to URL parsing according to the latest RFCs (and sanity) but unfortunately wasn't discovered until after 1.0 so could not be changed safely.

**Comment by smarterclayton** [comment]
I talked to @deads2k offline and I don't want us checking this generically because eventually someone *will* need a name with slashes and percents, and it's up to them to fix their client and server code (until Go delivers a resolution or we need to support it in rest handler)

**Comment by liggitt** [comment]
Just keep in mind that if a name with a slash results in an etcd key with a %2F, etcd will choke on that as well

Same goes for names like "." and ".." that get normalized out of URL paths

**Comment by smarterclayton** [comment]
Someone can implement a URL and underlying encoding scheme on top of urls.

**Comment by liggitt** [comment]
That's true for transforms into etcd keys, though "." and ".." wouldn't be addressable via the API

File: pkg/api/validation/validation.go (hunk 3 above)
Comment (rjnagal) [inline pkg/api/validation/validation.go]: this is fixed. needs a rebase.

**Comment by smarterclayton** [comment]
I'm going to say based on everything that consumers of Kube code should define a ValidateName that doesn't allow % or / (which most names shouldn't contain anyway) and that this is not necessary.

**Pull request closed without merging**